### ISSUE TITLE
How to Save the rendered form in database using ajax??

### ISSUE THREAD

**Issue by ArditKurti** [kevinchappell/formBuilder#367]

**Comment by ankurk91** [comment]
Here is the example -
* Create a html form
```html
<form id="my-form" >
<div id="form-render-wrap">
</div>

<button type="submit">Save</button>
</form>
```
* Add js code
```js
// using jQuery
$('#form-render-wrap').formRender(
    {
      formData: questionsJSON, // This is data you stored in database when you build the form
      dataType: 'json'
    }
);

$('#my-form').on('submit', function (e) {
  e.preventDefault();
// Collect form data
  var formData = $(this).serialize();
// Send data to server
  $.post('save-form.php', formData)
      .done(function (response) {
        console.log('form saved')
      }).fail(function (jqXHR) {
    console.log('form was not saved')
  });
  // Prevent form submission
  return false;
});
```
// Get posted data on server
```php
// save-form.php
print_r($_POST);
```
Hope this helps.

**Comment by abhimalik96** [comment]
hey ankurk91, thanks for then answer. I am very new to all this, I was hoping you could shed some more light on this solution. From what I have understood my JSON data is present in a javascript variable in the demo.js file. Now i just need to send this over to my node server in the backend,I would sincerely appreciate any guidance or advise

**Comment by ankurk91** [comment]
@abhimalik96 
I have no idea of your node-server backend, sorry

**Comment by Jojoshua** [comment]
Please review https://github.com/kevinchappell/formBuilder/pull/750

**Comment by sharmaanand086** [comment]
hello ankur ,
here is my code and error not able to submit to database please let me know as soon as possbile

![image](https://user-images.githubusercontent.com/38725515/43000633-e2051c64-8c3f-11e8-88f5-78d26e70ac58.png)

jQuery(function($) {
  var $fbEditor = $(document.getElementById('fb-editor')),
    $formContainer = $(document.getElementById('fb-rendered-form')),
    fbOptions = {
      onSave: function() {       
        $fbEditor.toggle();
        $formContainer.toggle();
        $('form', $formContainer).formRender({           
          formData: formBuilder.formData
        });
      }
    },
    formBuilder = $fbEditor.formBuilder(fbOptions);
  
$('#form-render-wrap',).formRender({        
          formData: formBuilder,
          dataType:'JSON'  
      }); 
$('#myform').submit(function () {
  var data = new FormData($('#myform')[0]);
  $.ajax({
      type: 'POST',
      url: 'formbuilder/db.php',
      data: data,
      contentType: false,
      processData: false,
      cache: false
  });
});

});

**Comment by sharmaanand086** [comment]
@ankurk91 
hey ankur  I am saving jquery form builder json content to database, how can i provide edit functionality for already saved content so that i can change the field properties and add new fields to the existing form.

**Comment by ankurk91** [comment]
@sharmaanand086
Please take a close look into my code example above, you will get answer.

**Comment by sharmaanand086** [comment]
@ankurk91 you have given the solution for saving form in json that i did but i want that form again edit form database am doing but not getting form at the place of drag and drop area 

<script type="text/javascript">
 
 jQuery(function($) {
  var $fbEditor = $(document.getElementById('fb-editor')),
    $formContainer = $(document.getElementById('fb-rendered-form'));
    
    fbOptions = {
      onSave: function() {
        formData ='<?php echo $formfile; ?>'; 
        $fbEditor.toggle();
        $formContainer.toggle();        
        $('form', $formContainer).formRender({        	
          formData: formData
        });
      }
    },
    formBuilder = $fbEditor.formBuilder(fbOptions);

  $('.edit-form', $formContainer).click(function() {
    $fbEditor.toggle();
    $formContainer.toggle();
  });
});

</script>

 
<?php 
$servername = "localhost";
$username = "root";
$password = "";
$dbname = "forms";
$conn = new mysqli($servername,$username,$password,$dbname);
if($conn->connect_error){
 echo $conn->connect_error;
} 
 $id= $_GET["id"];
// echo $id;
 $sql = "SELECT id, fdata  FROM form_element where id='$id' ";
$result = mysqli_query($conn, $sql); 
$formfile;
if ($result->num_rows > 0) {     
    // output data of each row
    while($row = $result->fetch_assoc()) {        
        $formfile = preg_replace('!\\r?\\n!', "", $row["fdata"]);   
        // echo $formfile;
    }
    
} else {
    echo "0 results";
}
 
 ?>

**Comment by ankurk91** [comment]
You should be using same form for creating and updating.
The `questionsJSON` is variable that holds the previously saved JSON 

```js
// using jQuery
$('#form-render-wrap').formRender(
    {
      formData: questionsJSON, // This is data you stored in database when you build the form
      dataType: 'json'
    }
);
```

**Comment by ankurk91** [comment]
You should be using [onSave](https://github.com/kevinchappell/formBuilder/blob/36d4d993252e172d6d77c562564574ed0daec96d/src/js/config.js#L57) callback to send data via ajax.

It seems like my example above is outdated.

**Comment by sharmaanand086** [comment]
@ankurk91  No sir am not getting onsave callback my form data at drag and drop area 

jquery-----------------------
var fbOptions = {     
    onSave: function(e, formData) {
      toggleEdit();
      formData ='<?php echo $formfile; ?>'; 
      $('#form-render-wrap').formRender({
        formData: formData,
       dataType: 'json'
      });
      var abc=formData;
     $.ajax({ 
              method:"POST",
              url: "",
              data:{abc:formData},
             }).done(function (response) {
              console.log(response)
               //console.log('form saved')
              // console.log(formBuilder.actions.getData());   
              }).fail(function (jqXHR) {
                console.log('form was not saved')
            });return false;
    
    window.sessionStorage.setItem('formData', JSON.stringify(formData));
    },  
  };
  var formData = window.sessionStorage.getItem('formData');
  var editing = true;

  if (formData) {
    fbOptions.formData = JSON.parse(formData);
  }

**Comment by ralphmallari** [comment]
@sharmaanand086 @ankurk91 

Hi. May I ask how did you save the json data in mysql database? I'm sorry I'm new to all of this. I know the process of saving but I don't know how to do it.

**Comment by ankurk91** [comment]
@ralphmallari 
I have not used this library since past 1 year.
I have shared my example but seems like API has changed.
I can not longer help in this.

The official docs/examples may help.

**Comment by sahilr2050** [comment]
@ankurk91 Can you please help me in this way?

I have made a form using form builder and save the template as in this attached screenshot in the mysql database.

![image](https://user-images.githubusercontent.com/10128422/44830764-f8fcc680-ac40-11e8-9a42-e8ceccb2de82.png)

Also, I have successfully rendered the form on a web page.
```
<form action="/" method="POST">
<div id="renderMe"></div> <!-- Form will be render here -->
<input type="submit" name="submit" value="submit">
</form>
```

Now, What I am looking for is. When click on submit button it should save the form response along with the form template and a data added by the user. for example.

![image](https://user-images.githubusercontent.com/10128422/44830812-3bbe9e80-ac41-11e8-95b2-5583e2bab2a0.png)

Above image is Database Table which stores whole HTML form with added data by the user.

1> How do I store whole form template with added response by the user
2> How can I show it in admin with response given by user after storing it into database.

Thank you

**Comment by ralphmallari** [comment]
You should just do it like you're just editing the form. Just add the "value" key in your json data with the attribute of whatever the response of the user. See the example code below

` "value": "Response of the user" `

Then just display it like how you displayed your rendered form. Only now make the rendered form disabled so the admin can only view it.

**Comment by sahilr2050** [comment]
@ralphmallari  Thanks for the response. Can you please give me a small piece of code for better understanding instead of one line code.

Also, What parameter should I add to do this "Only now make the rendered form disabled so the admin can only view it."?

**Comment by ralphmallari** [comment]
To answer that can you show some screenshot of your codes in displaying your rendered form?
And about the "parameter" you should just put your whole form inside of a fieldset tag and put a disabled attribute, like this

`<form>
    <fieldset disabled>
        <input/>
        <input/>
        <input/>
    </fieldset>
</form>`

**Comment by ralphmallari** [comment]
To answer that can you show some screenshot of your codes in displaying your rendered form?
And about the "parameter" you should just put your whole form inside of a fieldset tag and put a disabled attribute, like this

    <form>
    <fieldset disabled>
        <input/>
        <input/>
        <input/>
    </fieldset>
    </form>

**Comment by sahilr2050** [comment]
@ralphmallari  Thanks again. I did it in some different way. To show the form with user information I will not put it into the form tag as admin don't require to submit the form. They just want to see it. 

Also, How can I save the files which are uploaded from the formbuilder file upload element?

**Comment by sahilr2050** [comment]
@ralphmallari  Thanks mate. All done.

**Comment by ralphmallari** [comment]
Can you tell me if my reply helped you or did you do something else?

**Comment by rmintri5** [comment]
@ralphmallari  how can I add value key in current JSON and can you help me on that?

**Comment by ralphmallari** [comment]
@rmintri5 Can you tell me what are you trying to accomplish?

**Comment by mourareaupro** [comment]
Could you tell me how you stored the full JSON from the rendered from ? because actualy i can only get the key:value with the rendered form. but i need to store the full JSON as i did from the form-builder.

here my code : 

`<div class="sf_admin_form">
    <div class="content">
        <form class="my-form" name="my-form" id="my-form" >
            <div id="<?php echo $page->getIdPage(); ?>" class="render-wrap"></div>
                    <script>
                        jQuery(function($) {
                            var data = <?php echo json_encode($page->getForm()); ?>;
                            var text = data.replace(/&quot;/g, '"');
                            text = text.replace('"\\"[', '[');
                            text = text.replace(']\\""', ']');

                            $('.render-wrap').formRender({
                                formData: JSON.parse(text),
                                dataType: 'json'
                            });
                        });
                    </script>
            </div>
            <button type="submit">Save</button>
        </form>
</div>`

`$('#my-form').on('submit', function (e) {
        e.preventDefault();
    // Collect form data
        var formData = $(this).serializeArray();



        var id = document.getElementsByClassName('render-wrap');
        var id_page = id[0].getAttribute('id');
        var id_ca = '<?php echo sfContext::getInstance()->getUser()->getAttribute('id_ca') ?>';

        url = window.location.href;
        url = url.substring(0, url.indexOf('.php', 0) + 4);

        window.sessionStorage.setItem('formData', JSON.stringify(formData));

        var data = JSON.stringify(formData);

    // Send data to server
        $.post(url + "/ajax/save_page_answers", {data : data, id_page : id_page , id_ca : id_ca})
            .done(function (response) {
                console.log('form saved')
            }).fail(function (jqXHR) {
            console.log('form was not saved')
        });
        // Prevent form submission
        return false;
    });`

**Comment by ralphmallari** [comment]
@mourareaupro 
Do something like this, use ajax to pass the whole json data to your php file
`//use the "formBuilder.actions.getData('json', true);"`
`//store that value in variable, that is the json right there. See the code below`

`var fullJSON = formBuilder.actions.getData('json', true);`
  
  `$.ajax({`
    `type: 'POST', `
    `url: '{your url to save the data to}',`
    `data: "jsondata = "+fullJSON,`
    `success: function () {`
    `alert('success')`
  `},`
    `error: function () {`
    `alert('fail');`
  `}`
`});`

On your backend php to get the data and save it,
`jsondata = $_POST["jsondata"];`
`//code to save the whole data`

**Comment by mourareaupro** [comment]
@ralphmallari thanks for your answer, i already do this in order to save the form that i created with the different elements. i stored this in the database and now with the code in the preview post i render the form for the user in order to get the reponses but i would like to get the full json form, not only the key:value that i get with .serializeArray()

**Comment by kevinchappell** [comment]
can use `onSave` option now to trigger the $.ajax

**Comment by pavee92** [comment]
Hi evryone , it would be useful if you share step by step to save the form in database as i m new to this. I have look for many solution but couldnt understand the step . This is what i did so far : 

my index.php file 

<script>
jQuery(function($) {
  var fbEditor = document.getElementById('build-wrap');
  var formBuilder = $(fbEditor).formBuilder();


  
  

});


const actionButtons = [
    {
      id: 'export_xml',
      className: 'btn btn-success',
      label: 'Json File',
      type: 'button',
      events: {
        click: () => {
          var form_fields = $('.build_wrap').formBuilder('getData','xml');
		  
		  $.ajax({
		  dataType: "xml",
			  url : 'http://localhost/formbuilder/src/demo/test.php',
			  type: 'POST',
			  data : form_fields,
			  success: function (response) {
				  alert ('T.T');
			  },
			  error: function(xhr){
				  
				  alert ("T.T :" + xhr.status + " " + xhr.statusText);
			  }
		  })
		},
	  },
	},
  ]
  </script>




<div id="build-wrap">
<div id="getData">

<form action="test.php" method="POST">
<button id="build_wrap">Edit Form</button></div></div>
</form>
</body>
</html>



test.php file : 

<?php session_start(); ?>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<?php include ("connection.php");

$form_data = $_POST['form_fields'];
$xml = simplexml_load_string($form_data);
$result=[];
if(isset($xml->fields) and isset($xml->fields->field)) {
	
foreach($xml->fields->fields as $record) {

$array=(array)$record;
$record = current($array);
$record['options'] = next ($array);
$result[] = $record;

}	
}
return $result;
?>




Any suggestion of improvement on this would be helpful for me . Thanks in advance

**Comment by enegyluvagyok** [comment]
Hello!

My workaround for this particular problem was something like that. (Laravel)

First of all, I made a model for the forms I would like to store, and made the necessary migrations. 
Next I built a form element where I put the div that will contain the formbuilder itself. 
Magic comes just after this. (The html section in the documentation refers to something like that actually.)


  const form = $('#form-render-wrap').formBuilder(options);
  const escapeEl = document.createElement("textarea");
  const code = document.getElementById("markup");
  const $markup = $("<div/>");

  $('#myButton').on('click', function(e) {
    const formData = form.formData;
    const addLineBreaks = html => html;
    $markup.formRender({
      formData
    });
    code.innerText = addLineBreaks($markup.formRender("html"));
    var formHtml = code.innerText
    return false;
  });

This snippet will first print the json value of the formdata to the page in this element (include this in your html section):
<pre><p id="markup" style="display: none;"></p></pre>
You can hide it of course.  Once you press the button with id=#myButton you'll have the formbuilder form in html format saved to a javascript variable. Then you should make an ajax post request to store it in the database and use anywhere else in your application.

Hope this helps!
